Allow weather updates each hour even if weather hasn't changed

bradders2000 opened the pull request:
Even if the weather hasn't changed, the DB should get updated each hour so we know it's up-to-date for the current hour.

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
If the weather doesn't change from hour to hour, the DB entry looks 'old'. This code change makes it update the DB if it hasn't done so for the current hour. Map frontends can then tell if the weather is out of date and display accordingly.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I've been running this code change for ages now.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--  NOTE: In order to check code style locally and avoid having your build rejected by GitHub Actions, -->
<!--  run the following commands before you commit: `swiftlint` and `eslint`. Fix anyissues they point out. -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

123FLO321 commented on Sources/RealDeviceMap/Modell/Weather.swift:
```diff
@@ -437,9 +439,31 @@ class Weather: JSONConvertibleObject, WebHookEvent {
+        let oldUpdated = old.updated ?? 0
+        let newUpdated = UInt32(Date().timeIntervalSince1970)
+
+        hasChanged = ((newUpdated - (newUpdated % 3600)) > (oldUpdated - (oldUpdated % 3600)))
```
I am not sure i understand your logic here.
Would it be sufficient, if you just add `Date().timeIntervalSince1970 - (old.updated  ?? 0) > 3600` to the original or?

bradders2000 replied:
Isn't that just within the last hour, not since the top of the hour?

I had originally added it to the existing or, but swift complained about it being too complicated or something. Sounded like nonsense to me, but it wasn't happy. Had to break it down to how I have it now.

Fabio1988 replied:
so you calculate the value for top of the hour for newUpdated and oldUpdated and you compare it... RDM will always calculate until line 460, also when weather is not changing during hour?! :)

Fabio1988 commented:
Maybe we could add also a logic for frontend weather views... weather is displayed as red (not updated) after a certain amount of time.. But weather is fine until e.g. 3:59.... so it should be displayed in green?!

bradders2000 commented:
Yes, that's how my RDM frontend is configured to work and it's how my pr for pmsf works. 
This backend tweak is required to make both changes work nicely.

Fabio1988 commented:
@bradders2000 can you please rebase on master?
Would test this setup when Flo is starting deplyoment manually afterwards :)

Fabio1988 commented:
should we also update [index.js.mustache](resources/webroot/index.js.mustache) ?

`function getWeatherStyle (weather, ts) {
  const ago = ts - weather.updated
  let value
  if (ago <= 1800) {
    value = 0
  } else {
    value = Math.min((ago - 1800) / 1800, 1)
  }
  const hue = ((1 - value) * 120).toString(10)
  return { fillColor: ['hsl(', hue, ',100%,50%)'].join(''), color: 'black', opacity: 0.75, fillOpacity: 0.5, weight: 1.0 }
}`

bradders2000 commented:
> should we also update [index.js.mustache](resources/webroot/index.js.mustache) ? 150 seconds to check weather for update is a small number, weather is actuall for 3600 seconds, normally.
> 
> > function getWeatherStyle (weather, ts) {
> > const ago = ts - weather.updated
> > let value
> > if (ago <= 1800) {
> > value = 0
> > } else {
> > value = Math.min((ago - 1800) / 1800, 1)
> > }
> > const hue = ((1 - value) * 120).toString(10)
> > return { fillColor: ['hsl(', hue, ',100%,50%)'].join(''), color: 'black', opacity: 0.75, fillOpacity: 0.5, weight: 1.0 }
> > }

This PR is purely backend.

For the front end, I use this...

`function getWeatherStyle (weather, ts) {
  let updatedHour = weather.updated - (weather.updated % 3600)
  let thisHour = ts - (ts % 3600)
  let value
  if (updatedHour >= thisHour) {
     value = 0
  } else {
      value = 1
  }
  const hue = ((1 - value) * 120).toString(10)
  return { fillColor: ['hsl(', hue, ',100%,50%)'].join(''), color: 'black', opacity: 0.75, fillOpacity: 0.5, weight: 1.0 }
}
`
